sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/datasketchesAggregates.scala

```diff
@@ -265,6 +288,26 @@ case class HllUnionAgg(
   override protected def withNewChildrenInternal(newLeft: Expression, newRight: Expression):
   HllUnionAgg = copy(left = newLeft, right = newRight)
 
+  // Overrides for ExpectsInputTypes
+
+  override def checkInputDataTypes(): TypeCheckResult = {
```

Comment:
this looks mostly duplicated with L104-120 above. Can we dedup into one place and just call the helper twice instead?

Reply:
I think it's possible, but might be more valuable to create a new trait that extends ExpectsInputTypes that by default allows all of the types defined by inputTypes to be non-foldable, but provides a mechanism to override each type and define whether its foldable/non-foldable. Feels like this could be utilized by other functions too - thoughts?

Reply:
Could we define e.g.

```
  private def checkInputDataTypes(
      defaultCheck: TypeCheckResult, right: Expression): TypeCheckResult = {
    if (defaultCheck.isFailure) {
      defaultCheck
    } else if (!right.foldable) {
      DataTypeMismatch(
        errorSubClass = "NON_FOLDABLE_INPUT",
        messageParameters = Map(
          "inputName" -> "lgConfigK",
          "inputType" -> toSQLType(right.dataType),
          "inputExpr" -> toSQLExpr(right)
        )
      )
    } else {
      TypeCheckSuccess
    }
  }
```